File: element-recursion/josephus.js
/*
$ /usr/local/src/v8/d8 josephus.js
26
element recursion
7.496
7.517
7.522
7.527
7.526
7.524
7.526
7.527
7.524
7.526
*/


function countoff(soldiers, kth, k) {
   var i = soldiers.shift()

   if (soldiers.length > 0){
      if (k != kth){
         k++
         soldiers.push(i)
      } else {
         k = 1
      }
      return countoff(soldiers, kth, k)

   } else {
      return i
   }
}


function countoffSoldiers(n, kth) {
   var soldiers = []
   for (var i = 0; i < n; i++) soldiers[i] = i+1

   //return countoff(soldiers, kth, 1); // standard Josephus problem
   return countoff(soldiers, kth, kth)
}


//-----------------------------------------------------------------------------

function run_iterations(iterations, times) {
    for (var t = 0 ; t < times ; t++) {
        var start = new Date()
        for(var i = 0 ; i < iterations ; i++) {
            countoffSoldiers(40,3)
        }
        var end = new Date()
        print ((end.getTime() - start.getTime()) * 1000 / ITERS)
    }
}

print(countoffSoldiers(40,3))
ITERS = 1000000
print ("element recursion")
run_iterations(ITERS,10)

File: element-recursion/josephus.py
"""
$ python josephus.py
26
element recursion
43.7426130772
43.728605032
43.7135281563
43.7130749226
43.7156639099
43.7254951
43.7201170921
43.6938109398
43.693862915
43.6937048435
"""

from collections import deque

def countoff(soldiers, kth, k):
   i = soldiers.popleft()

   if soldiers:
      if k != kth:
         k += 1
         soldiers.append(i)
      else:
         k = 1

      return countoff(soldiers, kth, k)

   else:
      return i


def countoffSoldiers(n, kth):
   soldiers = deque( range(1,n+1) )

   #return countoff(soldiers, kth, 1) # standard Josephus problem
   return countoff(soldiers, kth, kth)


#-----------------------------------------------------------------------------

import time
def run_iterations(iterations, times) :
    print "element recursion" 
    for t in range(times) :
        start = time.time()
        for i in range(iterations) :
            countoffSoldiers(40,3)
        end = time.time()
        print ((end - start) * 1000000 / iterations)

print countoffSoldiers(40,3)
ITER = 1000000
run_iterations(ITER,10)

